src/mongo/db/audit/audit_commands.cpp

```diff
@@ -73,19 +74,37 @@ namespace mongo {
                    "Example: { logApplicationMessage: \"it's a trap!\" }";
         }
 
-        virtual void addRequiredPrivileges(const std::string& dbname,
-                                           const BSONObj& cmdObj,
-                                           std::vector<Privilege>* out) const override {
-            ActionSet actions;
-            actions.addAction(ActionType::logApplicationMessage);
-
-            // TODO: Investigate if using the 'normal resource'
-            // pattern of the new ResourcePattern API matches our
-            // original use of SERVER_RESOURCE_NAME. We may want
-            // to use somethine scoped to the given database name.
-            out->push_back(Privilege(ResourcePattern::forAnyNormalResource(), actions));
+        // Our original implementation was based on 'logApplicationMessage' action type
+        // then we realized this command should also work when 'applicationMessage' action type
+        // is granted.
+        // To keep compatibility for users of 'logApplicationMessage' we need to override
+        // 'checkAuthForCommand' because its default implementation only allows AND predicate
+        // for set of action types. We need OR here.
+        virtual Status checkAuthForCommand(Client* client,
+                                           const std::string& dbname,
+                                           const BSONObj& cmdObj) const override {
+            auto authzSess = AuthorizationSession::get(client);
+            if (authzSess->isAuthorizedForPrivilege(Privilege{ResourcePattern::forAnyNormalResource(), ActionType::logApplicationMessage}) ||
+                authzSess->isAuthorizedForPrivilege(Privilege{ResourcePattern::forAnyNormalResource(), ActionType::applicationMessage}))
+                return Status::OK();
+            return Status(ErrorCodes::Unauthorized, "unauthorized");
         }
 
+        // If we will ever remove 'logApplicationMessage' action type
+        // we can return to this implementation instead of 'checkAuthForCommand' override
+        //virtual void addRequiredPrivileges(const std::string& dbname,
+        //                                   const BSONObj& cmdObj,
+        //                                   std::vector<Privilege>* out) const override {
+        //    ActionSet actions;
+        //    actions.addAction(ActionType::logApplicationMessage);
```

Comment:
Hi Igor.
Shouldn't this commented-out line for a simpler future version of the method be for applicationMessage instead of logApplicationMessage?
```suggestion
        //    actions.addAction(ActionType::applicationMessage);
```

Reply:
fixed